File: challenges/03_code_to_comment/javascript/code_explainer.js
/**
 * Code Explainer Example
 * 
 * This file demonstrates GitHub Copilot's ability to help document code.
 * Each method contains complex algorithms that need documentation.
 * Use Copilot to help write detailed comments explaining what the code does.
 */

class TreeNode {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class TreeAnalysis {
    constructor(height, nodes, isBalanced, isBST) {
        this.height = height;
        this.nodes = nodes;
        this.isBalanced = isBalanced;
        this.isBST = isBST;
    }
}

class CodeExplainer {
    /**
     * TODO: Use GitHub Copilot to write detailed comments explaining this binary tree analysis algorithm.
     * Consider:
     * - What properties does it analyze?
     * - How does it determine if a tree is balanced?
     * - How does it verify BST properties?
     * - What are the edge cases?
     * - What is the time and space complexity?
     */
    static analyzeBinaryTree(root) {
        if (!root) {
            return new TreeAnalysis(0, 0, true, true);
        }

        const leftAnalysis = this.analyzeBinaryTree(root.left);
        const rightAnalysis = this.analyzeBinaryTree(root.right);

        const height = Math.max(leftAnalysis.height, rightAnalysis.height) + 1;
        const nodes = leftAnalysis.nodes + rightAnalysis.nodes + 1;
        const isBalanced = Math.abs(leftAnalysis.height - rightAnalysis.height) <= 1 
            && leftAnalysis.isBalanced && rightAnalysis.isBalanced;
        
        const isBST = leftAnalysis.isBST && rightAnalysis.isBST
            && (!root.left || this.findMax(root.left).value < root.value)
            && (!root.right || this.findMin(root.right).value > root.value);

        return new TreeAnalysis(height, nodes, isBalanced, isBST);
    }

    /**
     * TODO: Use GitHub Copilot to write detailed comments explaining this string pattern matching algorithm.
     * Consider:
     * - What algorithm is being implemented?
     * - How does the pattern table work?
     * - What are the key steps in the process?
     * - What are the edge cases?
     * - What is the time and space complexity?
     */
    static findPatternKMP(text, pattern) {
        if (!pattern || !text || pattern.length > text.length) {
            return [];
        }

        const lps = new Array(pattern.length).fill(0);
        let len = 0;
        let i = 1;

        while (i < pattern.length) {
            if (pattern[i] === pattern[len]) {
                len++;
                lps[i] = len;
                i++;
            } else {
                if (len !== 0) {
                    len = lps[len - 1];
                } else {
                    lps[i] = 0;
                    i++;
                }
            }
        }

        const matches = [];
        i = 0;
        let j = 0;

        while (i < text.length) {
            if (pattern[j] === text[i]) {
                i++;
                j++;
            }
            if (j === pattern.length) {
                matches.push(i - j);
                j = lps[j - 1];
            } else if (i < text.length && pattern[j] !== text[i]) {
                if (j !== 0) {
                    j = lps[j - 1];
                } else {
                    i++;
                }
            }
        }

        return matches;
    }

    /**
     * TODO: Use GitHub Copilot to write detailed comments explaining this graph algorithm.
     * Consider:
     * - What algorithm is being implemented?
     * - How does it handle negative weights?
     * - What happens with negative cycles?
     * - What are the edge cases?
     * - What is the time and space complexity?
     */
    static floydWarshall(graph) {
        const V = graph.length;
        const dist = Array.from({ length: V }, () => 
            Array.from({ length: V }, () => Infinity)
        );

        for (let i = 0; i < V; i++) {
            for (let j = 0; j < V; j++) {
                dist[i][j] = graph[i][j];
            }
        }

        for (let k = 0; k < V; k++) {
            for (let i = 0; i < V; i++) {
                for (let j = 0; j < V; j++) {
                    if (dist[i][k] !== Infinity && dist[k][j] !== Infinity
                        && dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }

        // Check for negative cycles
        for (let i = 0; i < V; i++) {
            if (dist[i][i] < 0) {
                throw new Error("Graph contains negative weight cycle");
            }
        }

        return dist;
    }

    static findMax(node) {
        while (node.right) {
            node = node.right;
        }
        return node;
    }

    static findMin(node) {
        while (node.left) {
            node = node.left;
        }
        return node;
    }
}

// Example usage
const root = new TreeNode(5);
root.left = new TreeNode(3);
root.right = new TreeNode(7);
root.left.left = new TreeNode(1);
root.left.right = new TreeNode(4);
root.right.left = new TreeNode(6);
root.right.right = new TreeNode(8);

console.log('Tree Analysis:', CodeExplainer.analyzeBinaryTree(root));
console.log('Pattern Matches:', CodeExplainer.findPatternKMP('AABAACAADAABAAABAA', 'AABA'));

const graph = [
    [0, 3, Infinity, 5],
    [2, 0, Infinity, 4],
    [Infinity, 1, 0, Infinity],
    [Infinity, Infinity, 2, 0]
];
console.log('Shortest Paths:', CodeExplainer.floydWarshall(graph));

// Practice exercises:
// TODO: Use GitHub Copilot to document these algorithms
// 1. Red-Black Tree operations
// 2. A* pathfinding algorithm
// 3. Quick Sort with different pivots
// 4. Topological sort
// 5. Minimum spanning tree

module.exports = CodeExplainer; 

File: challenges/03_code_to_comment/python/code_explainer.py
"""
Code Explainer Example

This file demonstrates GitHub Copilot's ability to help document code.
Each method contains complex algorithms that need documentation.
Use Copilot to help write detailed comments explaining what the code does.
"""

from typing import List, Optional, Dict, Any, Union
from dataclasses import dataclass

@dataclass
class TreeNode:
    value: int
    left: Optional['TreeNode'] = None
    right: Optional['TreeNode'] = None

@dataclass
class TreeAnalysis:
    height: int
    nodes: int
    is_balanced: bool
    is_bst: bool

class CodeExplainer:
    """Collection of algorithms that need documentation"""

    @staticmethod
    def analyze_binary_tree(root: Optional[TreeNode]) -> TreeAnalysis:
        """
        TODO: Use GitHub Copilot to write detailed comments explaining this binary tree analysis algorithm.
        Consider:
        - What properties does it analyze?
        - How does it determine if a tree is balanced?
        - How does it verify BST properties?
        - What are the edge cases?
        - What is the time and space complexity?
        """
        if not root:
            return TreeAnalysis(0, 0, True, True)

        left_analysis = CodeExplainer.analyze_binary_tree(root.left)
        right_analysis = CodeExplainer.analyze_binary_tree(root.right)

        height = max(left_analysis.height, right_analysis.height) + 1
        nodes = left_analysis.nodes + right_analysis.nodes + 1
        is_balanced = abs(left_analysis.height - right_analysis.height) <= 1 \
            and left_analysis.is_balanced and right_analysis.is_balanced
        
        is_bst = left_analysis.is_bst and right_analysis.is_bst \
            and (not root.left or CodeExplainer.find_max(root.left).value < root.value) \
            and (not root.right or CodeExplainer.find_min(root.right).value > root.value)

        return TreeAnalysis(height, nodes, is_balanced, is_bst)

    @staticmethod
    def find_pattern_kmp(text: str, pattern: str) -> List[int]:
        """
        TODO: Use GitHub Copilot to write detailed comments explaining this string pattern matching algorithm.
        Consider:
        - What algorithm is being implemented?
        - How does the pattern table work?
        - What are the key steps in the process?
        - What are the edge cases?
        - What is the time and space complexity?
        """
        if not pattern or not text or len(pattern) > len(text):
            return []

        lps = [0] * len(pattern)
        length = 0
        i = 1

        while i < len(pattern):
            if pattern[i] == pattern[length]:
                length += 1
                lps[i] = length
                i += 1
            else:
                if length != 0:
                    length = lps[length - 1]
                else:
                    lps[i] = 0
                    i += 1

        matches = []
        i = 0
        j = 0

        while i < len(text):
            if pattern[j] == text[i]:
                i += 1
                j += 1
            if j == len(pattern):
                matches.append(i - j)
                j = lps[j - 1]
            elif i < len(text) and pattern[j] != text[i]:
                if j != 0:
                    j = lps[j - 1]
                else:
                    i += 1

        return matches

    @staticmethod
    def floyd_warshall(graph: List[List[float]]) -> List[List[float]]:
        """
        TODO: Use GitHub Copilot to write detailed comments explaining this graph algorithm.
        Consider:
        - What algorithm is being implemented?
        - How does it handle negative weights?
        - What happens with negative cycles?
        - What are the edge cases?
        - What is the time and space complexity?
        """
        V = len(graph)
        dist = [[float('inf')] * V for _ in range(V)]

        for i in range(V):
            for j in range(V):
                dist[i][j] = graph[i][j]

        for k in range(V):
            for i in range(V):
                for j in range(V):
                    if dist[i][k] != float('inf') and dist[k][j] != float('inf') \
                        and dist[i][k] + dist[k][j] < dist[i][j]:
                        dist[i][j] = dist[i][k] + dist[k][j]

        # Check for negative cycles
        for i in range(V):
            if dist[i][i] < 0:
                raise ValueError("Graph contains negative weight cycle")

        return dist

    @staticmethod
    def find_max(node: TreeNode) -> TreeNode:
        while node.right:
            node = node.right
        return node

    @staticmethod
    def find_min(node: TreeNode) -> TreeNode:
        while node.left:
            node = node.left
        return node

def run_examples():
    """Example usage of the algorithms"""
    # Create a binary search tree
    root = TreeNode(5)
    root.left = TreeNode(3)
    root.right = TreeNode(7)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(4)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(8)

    # Test tree analysis
    print('Tree Analysis:', CodeExplainer.analyze_binary_tree(root))

    # Test pattern matching
    print('Pattern Matches:', CodeExplainer.find_pattern_kmp('AABAACAADAABAAABAA', 'AABA'))

    # Test shortest paths
    graph = [
        [0, 3, float('inf'), 5],
        [2, 0, float('inf'), 4],
        [float('inf'), 1, 0, float('inf')],
        [float('inf'), float('inf'), 2, 0]
    ]
    print('Shortest Paths:', CodeExplainer.floyd_warshall(graph))

if __name__ == "__main__":
    run_examples()

# Practice exercises:
# TODO: Use GitHub Copilot to document these algorithms
# 1. Red-Black Tree operations
# 2. A* pathfinding algorithm
# 3. Quick Sort with different pivots
# 4. Topological sort
# 5. Minimum spanning tree 